src/common.rs

```diff
@@ -0,0 +1,264 @@
+        let expected_bytes = pixels_to_bytes::<C>(expected_pixels);
+
+        if buffer.len() < expected_bytes {
+            return Err(BufferError::TruncatedData {
+                expected_buffer_size: expected_bytes,
+            });
+        }
+
+        Ok(())
+    }
+
+    // TODO: ensure that size.width and size.height are small enough
+    pub const fn index(&self, point: Point) -> Result<usize, OutOfBoundsError> {
+        // point.x and point.y are implicitly also checked to be >= 0:
+        // Casting a negative i32 value to u32 will result in a value >= 0x80000000, which is
+        // larger thant the maximum allowed buffer size.
+        if point.x as u32 >= self.size.width || point.y as u32 >= self.size.height {
+            return Err(OutOfBoundsError);
+        }
+
+        Ok(match A::ARRANGEMENT {
+            PixelArrangementEnum::Horizontal => point.x as usize + point.y as usize * self.stride,
+            PixelArrangementEnum::Vertical => point.x as usize * self.stride + point.y as usize,
+        })
+    }
+}
+
+const fn default_stride<C: StorablePixelColor, A: PixelArrangement>(size: Size) -> usize {
+    let mut stride = match A::ARRANGEMENT {
+        PixelArrangementEnum::Horizontal => size.width as usize,
+        PixelArrangementEnum::Vertical => size.height as usize,
+    };
+
+    if C::Raw::BITS_PER_PIXEL < 8 {
+        let pixels_per_bit = 8 / C::Raw::BITS_PER_PIXEL;
+        stride = (stride + (pixels_per_bit - 1)) / pixels_per_bit * pixels_per_bit;
+    }
+
+    stride
+}
+
+/// Calculates the required buffer size.
+///
+/// This function is a workaround for current limitations in Rust const generics.  It can be used to
+/// calculate the `N` parameter for [`ArrayFramebuffer`] based on the size and color type of the
+/// framebuffer.
+///
+/// [`ArrayFramebuffer`]: crate::framebuffer::ArrayFramebuffer
+pub const fn buffer_size<C: StorablePixelColor, A: PixelArrangement>(size: Size) -> usize {
+    buffer_size_with_stride::<C, A>(size, default_stride::<C, A>(size))
+}
+
+/// Calculates the required buffer size with a custom stride.
+pub const fn buffer_size_with_stride<C: StorablePixelColor, A: PixelArrangement>(
+    size: Size,
+    stride: usize,
+) -> usize {
+    let pixels = match A::ARRANGEMENT {
+        PixelArrangementEnum::Horizontal => stride * size.height as usize,
+        PixelArrangementEnum::Vertical => stride * size.width as usize,
+    };
+
+    pixels_to_bytes::<C>(pixels)
+}
+
+/// Pixel arrangement.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "defmt", derive(::defmt::Format))]
+#[non_exhaustive]
+pub enum PixelArrangementEnum {
+    /// Horizontal arrangement.
+    Horizontal,
+    /// Vertical arrangement.
+    Vertical,
+}
+
+/// Pixel arrangement.
+pub trait PixelArrangement: private::Sealed + Copy {
+    /// Pixel arrangement.
+    const ARRANGEMENT: PixelArrangementEnum;
```

Comment:
I think this type could use a `fn xy_to_main_cross(T, T) -> (T, T)` (or variants like main_from_xy/cross_from_xy) function where one implementation returns the arguments straight, the other swaps them. I've seen plenty of switching on the enum value during my cursory first glance that could probably be cleaned up this way. That being said, the second to last commit was so large I skipped it, and I think that was where this const was updated from `IS_HORIZONTAL`, so I probably made the conclusion based on outdated info.

Reply:
I think I should have given an example:

```rust
let (m, c) = A::xy_to_main_cross(point.x as usize, point.y as usize);
m * self.stride + c
```

Reply:
OK, that is much better.

Reply:
Thank you :) Though this turned out to be fewer occurrances than I remembered.